File: projekty/PO_Projekt3/zwierze.py
from organizm import Organizm
import random

class Zwierze(Organizm):
    def __init__(self, swiat, polozenie, sila, inicjatywa, szansaRuchu, zasiegRuchu, szansaUcieczki, szansaRozmnazania):
        super().__init__(swiat, polozenie, sila, inicjatywa)
        self._szansaRuchu = szansaRuchu
        self._zasiegRuchu = zasiegRuchu
        self._szansaUcieczki = szansaUcieczki
        self._szansaRozmnazania = szansaRozmnazania

    def getSzansaRuchu(self):
        return self._szansaRuchu

    def getZasiegRuchu(self):
        return self._zasiegRuchu

    def getSzansaUcieczki(self):
        return self._szansaUcieczki

    def getSzansaRozmnazania(self):
        return self._szansaRozmnazania

    def akcja(self):
        if not self._zyje:
            return

        if not random.random() < self._szansaRuchu:
            return

        for i in range(self._zasiegRuchu):
            cel = self._swiat.losoweSasiedniePole(self._polozenie)

            if self._swiat.poleZajete(cel):
                self.kolizja(self._swiat.organizmNaPolu(cel), True)
                return

            self.przemiescSieNaPole(cel)

    def kolizja(self, przeciwnik, atak):
        if not atak:
            return

        if self.tegoSamegoTypu(przeciwnik):
            self.rozmnazanie(przeciwnik)
            return

        self.atakowanie(przeciwnik)

    def mozeUciec(self):
        return random.random() < self._szansaUcieczki

    def przemiescSieNaPole(self, polozenie):
        self._swiat.usunOrganizmZPola(self)
        self._polozenie = polozenie
        self._swiat.dodajOrganizmDoPola(self)

    def ucieknij(self, przeciwnik, atak):
        cele = self._swiat.sasiednieWolnePola(self._polozenie)

        if atak:
            if len(cele) == 0:
                return

            cele.append(self._polozenie)
            self.przemiescSieNaPole(self._swiat.losowePole(cele))
        else:
            if len(cele) == 0:
                self.zamienMiejscami(przeciwnik)
                return

            self.przemiescSieNaPole(self._swiat.losowePole(cele))

    def rozmnazanie(self, partner):
        if not self.mozeSieRozmnazac():
            return

        pola = self._swiat.sasiednieWolnePola(self._polozenie)
        polaPartnera = self._swiat.sasiednieWolnePola(partner._polozenie)
        pola.extend(polaPartnera)

        if len(pola) == 0:
            return

        dziecko = self.stworzDziecko(self._swiat.losowePole(pola))
        self._swiat.dodajDziecko(dziecko)

        self._swiat.dodajKomunikat(str(self._polozenie) + ": " + self.nazwaKlasy() + " ma dziecko")

    def mozeSieRozmnazac(self):
        return random.random() < self._szansaRozmnazania

    def atakowanie(self, przeciwnik):
        if self.mozeUciec():
            self.ucieknij(przeciwnik, True)
            self._swiat.dodajKomunikat(str(self._polozenie) + ": " + self.nazwaKlasy() + " ucieka przed " + przeciwnik.nazwaKlasy())
        elif przeciwnik.mozeUciec():
            przeciwnik.ucieknij(self, False)
            self._swiat.dodajKomunikat(str(self._polozenie) + ": " + przeciwnik.nazwaKlasy() + " ucieka przed " + self.nazwaKlasy())
        elif przeciwnik.odbilAtak(self):
            self._swiat.dodajKomunikat(str(self._polozenie) + ": " + przeciwnik.nazwaKlasy() + " odbil atak od " + self.nazwaKlasy())
            return
        elif self._sila >= przeciwnik._sila:
            przeciwnik.kolizja(self, False)
            przeciwnik.zakonczZycie()
            self.przemiescSieNaPole(przeciwnik._polozenie)
            self._swiat.dodajKomunikat(str(self._polozenie) + ": " + self.nazwaKlasy() + " zabija " + przeciwnik.nazwaKlasy())
        else:
            self._swiat.dodajKomunikat(str(self._polozenie) + ": " + przeciwnik.nazwaKlasy() + " broni sie przed " + self.nazwaKlasy())
            przeciwnik.kolizja(self, False)
            self.zakonczZycie()

File: projekty/PO_Projekt2/ProgOb_Projekt2/src/main/java/pl/Dybowski/Adrian/Projekt2/WirtualnySwiat/Zwierzeta/Zwierze.java
package pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Zwierzeta;

import java.util.HashSet;

import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Organizm;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.RodzajeKomunikatow;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Swiat;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Wspolrzedne2D;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Rosliny.Roslina;

public abstract class Zwierze extends Organizm {
	protected final double szansaRuchu;
	protected final int zasiegRuchu;
	protected final double szansaUcieczki;
	protected final double szansaRozmazania;
	
	public Zwierze(Swiat swiat, Wspolrzedne2D polozenie, int sila, int inicjatywa, 
					double szansaRuchu, int zasiegRuchu, double szansaUcieczki, double szansaRozmazania) {
		super(swiat, polozenie, inicjatywa, sila);
		this.szansaRuchu = szansaRuchu;
		this.zasiegRuchu = zasiegRuchu;
		this.szansaUcieczki = szansaUcieczki;
		this.szansaRozmazania = szansaRozmazania;
	}
	
	@Override
	public void akcja() {
		if (!zyje) return;
		
		if (!swiat.szansaTrafiona(szansaRuchu)) return;
		
		for (int i = 0; i < zasiegRuchu; i++) {
			Wspolrzedne2D cel = swiat.losoweSasiedniePole(polozenie);
			
			if (swiat.poleZajete(cel)) {
				kolizja(swiat.organizmNaPolu(cel), true);
				return;
			}
			
			przemiescSieNaPole(cel);
		}
	}
	
	@Override
	public void kolizja(Organizm przeciwnik, boolean atak) {
		if (!atak) return;
		
		if (tegoSamegoTypu(przeciwnik)) {
			rozmnazanie(przeciwnik);
			return;
		}
		atakowanie(przeciwnik);
	}
	
	@Override
	public boolean mozeUciec() {
		return swiat.szansaTrafiona(szansaUcieczki);
	}
	
	public void przemiescSieNaPole(Wspolrzedne2D polozenie) {
		swiat.usunOrganizmZPola(this);
		this.polozenie = polozenie;
		swiat.dodajOrganizmDoPola(this);
	}
	
	@Override
	public void ucieknij(Organizm przeciwnik, boolean atak) {
		HashSet<Wspolrzedne2D> cele = swiat.sasiednieWolnePola(this.polozenie);
		
		if (atak) {
			if (cele.size() == 0) return;
			
			cele.add(this.polozenie);	// organizm moze wrocic na pole z ktorego atakowal
			
			this.przemiescSieNaPole(swiat.losowePole(cele));
		}
		else {
			if (cele.size() == 0) {
				this.zamienMiejscami(przeciwnik);
				return;
			}
			
			this.przemiescSieNaPole(swiat.losowePole(cele));
		}
	}
	
	public void rozmnazanie(Organizm partner) {
		if (!mozeSieRozmnazac()) return;
		
		HashSet<Wspolrzedne2D> pola = swiat.sasiednieWolnePola(this.polozenie);
		HashSet<Wspolrzedne2D> polaPartnera = swiat.sasiednieWolnePola(this.polozenie);
		pola.addAll(polaPartnera);	// dziecko moze pojawic sie przy ktorymkolwiek z rodzicow
		
		if (pola.size() == 0) return;
		
		Organizm dziecko = this.stworzDziecko(swiat.losowePole(pola));
		swiat.dodajDziecko(dziecko);
		
		swiat.dodajKomunikat(dziecko, RodzajeKomunikatow.rodziSie);
	}
	
	public boolean mozeSieRozmnazac() {
		return swiat.szansaTrafiona(szansaRozmazania);
	}
	
	public void atakowanie(Organizm przeciwnik) {
		if (this.mozeUciec()) {
			this.ucieknij(przeciwnik, true);
			swiat.dodajKomunikat(przeciwnik, this, RodzajeKomunikatow.ucieka);
		}
		else if (przeciwnik.mozeUciec()) {
			przeciwnik.ucieknij(this, false);
			swiat.dodajKomunikat(this, przeciwnik, RodzajeKomunikatow.ucieka);
		}
		else if (przeciwnik.odbilAtak(this)) {
			swiat.dodajKomunikat(this, przeciwnik, RodzajeKomunikatow.odbija);
			return;
		}
		else if (this.sila >= przeciwnik.getSila()) {
			przeciwnik.kolizja(this, false);
			przeciwnik.zakonczZycie();
			this.przemiescSieNaPole(przeciwnik.getPolozenie());
			
			if (przeciwnik instanceof Roslina) {
				swiat.dodajKomunikat(this, przeciwnik, RodzajeKomunikatow.zjada);
			}
			else {
				swiat.dodajKomunikat(this, przeciwnik, RodzajeKomunikatow.zabija);
			}
		}
		else {
			przeciwnik.kolizja(this, false);
			this.zakonczZycie();
			

			if (przeciwnik instanceof Roslina) {
				swiat.dodajKomunikat(this, przeciwnik, RodzajeKomunikatow.zatruwaSie);
			}
			else {
				swiat.dodajKomunikat(this, przeciwnik, RodzajeKomunikatow.broniSie);
			}
		}
	}

}
